Hello World Precompile tutorial only (#261)
ava-labs/subnet-evm: cmd/README.md

# Precompile Generation Tutorial

> We can now generate a stateful precompile with the Precompile gen tool!

Reviewer: Can we improve this introduction to briefly explain what a stateful precompile is, what the precompilegen tool does, and what users will do in this tutorial?

Reviewer: We should make it clear that this will be a walk through of creating a simple HelloWorld stateful precompile

Author: Ah sounds good

### Assumption of Knowledge

> Before starting this tutorial it would be helpful if you had some context on the EVM, precompiles, and stateful precompiles. 
> Here are some resources to get started put together. 

Reviewer: ```suggestion

Before building a stateful precompile, you will want to make sure that you have an understanding of the EVM, precompiles, and stateful precompiles. 

Here are some resources to get started put together. 
```

Author: thanks

- [The Ethereum Virtual Machine](https://github.com/ethereumbook/ethereumbook/blob/develop/13evm.asciidoc)
- [Precompiles in Solidity](https://medium.com/@rbkhmrcr/precompiles-solidity-e5d29bd428c4)
- [Deconstructing a Smart Contract](https://blog.openzeppelin.com/deconstructing-a-solidity-contract-part-i-introduction-832efd2d7737/)
- [Layout of State Variables in Storage](https://docs.soliditylang.org/en/v0.8.10/internals/layout_in_storage.html)
- [Layout in Memory](https://docs.soliditylang.org/en/v0.8.10/internals/layout_in_memory.html)
- [Layout of Call Data](https://docs.soliditylang.org/en/v0.8.10/internals/layout_in_calldata.html)
- [Contract ABI Specification](https://docs.soliditylang.org/en/v0.8.10/abi-spec.html)
- [Customizing the EVM with Stateful Precompiles](https://medium.com/avalancheavax/customizing-the-evm-with-stateful-precompiles-f44a34f39efd)
- [Precompiles in Solidity](https://medium.com/@rbkhmrcr/precompiles-solidity-e5d29bd428c4)

> - [Customizing the EVM with Stateful Precompiles](https://medium.com/avalancheavax/customizing-the-evm-with-stateful-precompiles-f44a34f39efd)

Reviewer: These are both duplicates

Author: thanks

 
## Tutorial

> Let's start by creating the Solidity interface that we want to implement. We can put this  in `./contract-examples/contracts`

Reviewer: ```suggestion
Let's start by creating the Solidity interface that we want our precompile to implement. We will place this file in `./contract-examples/contracts`
```

```
// (c) 2022-2023, Ava Labs, Inc. All rights reserved.
// See the file LICENSE for licensing terms.

// SPDX-License-Identifier: MIT

pragma solidity >=0.8.0;

interface IHelloWorld {
 // sayHello returns the string located at [key]
  function sayHello() external returns (string calldata);

// setGreeting sets the string located at [key]
  function setGreeting(string calldata response) external;
}
```

Now we have an interface that our precompile can implement!
Let's create an [abi](https://docs.soliditylang.org/en/v0.8.13/abi-spec.html#:~:text=Contract%20ABI%20Specification-,Basic%20Design,as%20described%20in%20this%20specification.) of our solidity code.

In the same `./contract-examples/contracts` directory, let's download solc and run
```
brew update
brew tap ethereum/ethereum
brew install solidity

> 

Reviewer: Can we switch to providing a link and telling the user to install solidity here instead? Not everyone will be able to install via homebrew, so we should link to the official documentation and ask users to ensure they have correctly installed it.

We can switch to asking users to run a basic solc command to get the version and confirm that they have installed it correctly rather than specifying the exact instructions here.

solc --abi IHelloWorld.sol -o .
```

This spits out the abi code. Let's move it into a brand new folder in 

> `./contract-examples/contracts/contract-abis`. 

Reviewer: Can we ask users to keep the ABI file in the same folder instead of creating a new one ie. `./contract-examples/contracts/` ?

IHelloWorld.abi

```
[{"inputs":[],"name":"sayHello","outputs":[{"internalType":"string","name":"","type":"string"}],"stateMutability":"nonpayable","type":"function"},{"inputs":[{"internalType":"string","name":"recipient","type":"string"}],"name":"setGreeting","outputs":[],"stateMutability":"nonpayable","type":"function"}]
```

## Precompile tool

The precompile tool can take in 4 arguments. 

### `--abi` (string):

It needs an ABI input so it can bind it to the precompile template.

`--abi ./contract-examples/contracts/contract-abis/IHelloWorld.abi`

### `--type `(string):

It takes in a type which it uses as a struct name for the precompile. This is optional 
and will default to the abi name. 

`--type HelloWorld`

### `--pkg` (string):

It also takes in a pkg, which is the package name to generate the precompile into. 
This is optional and it defaults to "precompile".

`--pkg precompile` 

### `--out `(string):

Finally it can take in an out which is the path and name of the output file of the generated precompile. 

> 

Reviewer: We should separate this documentation from the tutorial itself, so that there is the documentation of what the tool does and a separate tutorial that walks the user through how to use it.

Author: I find this helpful here, because it allows for the user on how to customize their command. Can we keep it here for now?

Reviewer 2: I agree, if we aim to put this into docs repo we should split this and put PrecompileGen tool under Tools section in https://docs.avax.network/subnets. Then we can put a link here instead.

Alternatively for now we can (and probably should) move this section under `cmd/precompilegen/README.md`, and link here.

Reviewer: Let's put this into the docs repo and link to it or give a full command with the fill in the blanks of the command arguments here, a tutorial is not the best place to provide reference documentation like this. I actually found @dasconnor 's recent tweet to be great on this point: https://documentation.divio.com/introduction/

func (c *sha256hash) RequiredGas(input []byte) uint64 {
	return uint64(len(input)+31)/32*params.Sha256PerWordGas + params.Sha256BaseGas
}
```

We'll probably have to come back later and change these costs. 

## Step 3: Add Custom Code

Ok time to `CTRL F` throughout the file with `CUSTOM CODE STARTS HERE` to find the areas in the precompile that we need to modify. 

![](2022-09-01-22-48-26.png)

If we use any of the following imports we can remove the reference imports.
We can come back if needed to remove. Next we see this in `Equals()`.

![](2022-09-01-22-51-30.png)

We can skip this step since our HelloWorldConfig struct looks like this.

```
// HelloWorldConfig implements the StatefulPrecompileConfig
// interface while adding in the HelloWorld specific precompile address.
type HelloWorldConfig struct {
	UpgradeableConfig
}
```

**Optional Note** 

If our IHelloWorldConfig wrapped another config in its struct to implement the StatefulPrecompileConfig
like so 

```
// IHelloWorldConfig implements the StatefulPrecompileConfig
// interface while adding in the IHelloWorld specific precompile address.
type IHelloWorldConfig struct {
	UpgradeableConfig

>     AllowListConfig

Reviewer: Can we align these two fields?

Author: thanks

Author: Awkward indents here fix the real view

}
```

> We would have modify the `Equal()` function as follows: 

Reviewer: Can we fix the grammar here?

Author: thanks

	// Do some processing and pack the output
	packedOutput, err := PackSayHelloOutput(string(common.TrimLeftZeroes(value.Bytes())))
	if err != nil {
		return nil, remainingGas, err
	}

	// Return the packed output and the remaining gas
	return packedOutput, remainingGas, nil
}
```

Finally we can modify our `setGreeting()` function 

```
func setGreeting(accessibleState PrecompileAccessibleState, caller common.Address, addr common.Address, input []byte, suppliedGas uint64, readOnly bool) (ret []byte, remainingGas uint64, err error) {
	if remainingGas, err = deductGas(suppliedGas, SetGreetingGasCost); err != nil {
		return nil, 0, err
	}
	if readOnly {
		return nil, remainingGas, vmerrs.ErrWriteProtection
	}
	// attempts to unpack [input] into the arguments to the SetGreetingInput.
	// Assumes that [input] does not include selector
	// You can use unpacked [inputStruct] variable in your code
	inputStr, err := UnpackSetGreetingInput(input)
	if err != nil {
		return nil, remainingGas, err
	}

	// CUSTOM CODE STARTS HERE
	// setGreeting is the execution function of "SetGreeting(name string)" and sets the storageKey in the string returned by hello world

> 	res := common.LeftPadBytes([]byte(inputStr), common.HashLength)

Reviewer: one note: if the string is longer than 32 bytes, then it will end up getting cropped from the left here, so we can either leave this as is or add a check for the length of the string to make sure it returns an error instead if the string will be cropped.

Author: Ah ok

# Precompile Generation Tutorial

> A stateful precompile allows us to add more functionality and customization to the EVM. A stateful precompile builds on a precompile in that it adds state access. This means our precompile can manage state as well as interacting with EVM state. 

Reviewer: Can we improve the phrasing here so we don't start two consecutive sentences in the same way?

> A stateful precompile follows this interface. 

Reviewer: This one follows the same format as well and doesn't provide much information. What is the purpose of this interface?

Reviewer: Bumping this comment

``` go
// StatefulPrecompiledContract is the interface for executing a precompiled contract
type StatefulPrecompiledContract interface {
	// Run executes the precompiled contract.
	Run(accessibleState PrecompileAccessibleState, caller common.Address, addr common.Address, input []byte, suppliedGas uint64, readOnly bool) (ret []byte, remainingGas uint64, err error)
}

```

> It looks like we have to create an actual precompile, an interface for the precompile, a contract that interacts with the precompile, tests for that contract, and small modifications throughout the EVM to enable the precompile. This seems like a lot of work. Luckily, we have the precompile generation tool! 

Reviewer: Can we start this a different way? "It looks like" sounds wishy washy

> This tool does a bulk of the precompile creation work in one command. In this tutorial, we will be walking through how to create a working precompile from scratch. 

Reviewer: ```suggestion
The precompilegen tool will auto-generate the function stubs for a specified Solidity interface that we want our precompile to implement. In this tutorial, we will walk through the steps necessary to fill out the auto-generated precompile stub and implement a simple precompile.
```

### Assumption of Knowledge

> Before building a stateful precompile, you will want to make sure that you have an understanding of the EVM, precompiles, and stateful precompiles. 

Reviewer: ```suggestion

Before building a stateful precompile, you will want to make sure that you have an understanding of the EVM, precompiles, and stateful precompiles. 
```

> Here are some resources to get started put together. 

Reviewer: ```suggestion
Here are some helpful resources on the EVM to get started: 
```

- [The Ethereum Virtual Machine](https://github.com/ethereumbook/ethereumbook/blob/develop/13evm.asciidoc)
- [Precompiles in Solidity](https://medium.com/@rbkhmrcr/precompiles-solidity-e5d29bd428c4)
- [Deconstructing a Smart Contract](https://blog.openzeppelin.com/deconstructing-a-solidity-contract-part-i-introduction-832efd2d7737/)
- [Layout of State Variables in Storage](https://docs.soliditylang.org/en/v0.8.10/internals/layout_in_storage.html)
- [Layout in Memory](https://docs.soliditylang.org/en/v0.8.10/internals/layout_in_memory.html)
- [Layout of Call Data](https://docs.soliditylang.org/en/v0.8.10/internals/layout_in_calldata.html)
- [Contract ABI Specification](https://docs.soliditylang.org/en/v0.8.10/abi-spec.html)
- [Precompiles in Solidity](https://medium.com/@rbkhmrcr/precompiles-solidity-e5d29bd428c4)
- [Customizing the EVM with Stateful Precompiles](https://medium.com/avalancheavax/customizing-the-evm-with-stateful-precompiles-f44a34f39efd)

 
## Tutorial

> Let's start by creating the Solidity interface that we want our precompile to implement. We will place this file in `./contract-examples/contracts`

Reviewer: Can we add more detail here so that it's clear what the interface will do. We should say something about implementing a Hello World interface before laying out the solidity.

# Stateful Precompile Generation Tutorial

> In this tutorial,  we are going to walkthrough how we can generate a stateful precompile from scratch. Before we start, let's brush up on what a precompile is, what a stateful precompile is, and why this is extremely useful. 

Author: thanks

> ## Precompiled Contracts

Author: thanks

Precompiles were introduced to Ethereum as a way to solve the problem of allowing complex cryptographic computations to be usable in the EVM without having to deal with EVM overhead. The following precompiles are currently included: ecrecover, sha256, blake2f, ripemd-160, Bn256Add, Bn256Mul, Bn256Pairing, the identity function, and modular exponentiation.

> We can see these precompile mappings from address to function here in the ethereum vm. 

Author: thanks

```

Smart contracts in solidity are compiled and converted into bytecode when they are first deployed. They are then stored on the blockchain and an address (usually known as the contract address) is assigned to it. When a user calls a function from a smart contract, it goes through the `CALL` function in the EVM. It takes in the caller address, the contract address, the input (function’s signature (truncated to the first leading four bytes) followed by the packed arguments data), gas, and value (native token). The function selector from the input lets the EVM know where to start from in the bytecode of the smart contract. It then executes a series of instructions (EVM opcodes) and returns the result. 

When a precompile function is called, it still goes through the `CALL` function in the EVM. However, it works a little differently. The EVM checks if the address is a precompile address from the mapping list and if so redirects to the precompile function. 

``` go
  if p := precompiles[addr]; p != nil {
    return RunPrecompiledContract(p, input, contract)
  }
```

> The evm then performs the function and subtracts the `RequiredGas`.

Author: thanks

Precompiles provide complex library functions that are commonly used in smart contracts and do not use EVM opcodes which makes execution faster and gas costs lower.

## Stateful Precompiled Contracts

> A stateful precompile allows us to add even more functionality and customization to the EVM. It builds on a precompile in that it adds state access. 

Reviewer 2: We should emphasize that stateful precompiled contracts are not available in default EVM, but something we added for Coreth/Subnet-EVM

Author: thanks

A stateful precompile follows this interface. 
``` go
// StatefulPrecompiledContract is the interface for executing a precompiled contract
type StatefulPrecompiledContract interface {
	// Run executes the precompiled contract.
	Run(accessibleState PrecompileAccessibleState, caller common.Address, addr  common.Address, input []byte, suppliedGas uint64, readOnly bool) (ret []byte, remainingGas uint64, err error)

}
```

Notice the most important difference between the stateful precompile and precompile interface. We now inject state access to the `Run` function. Precompiles only took in a single byte slice as input. However, stateful precompile functions have complete access to the EVM state, and can be used to implement a much wider range of functionalities.

> ## Why This is Useful

Reviewer 2: actually seems we don't need this to be an additional section. we can move the content under Stateful Precompiled Contracts

Author: thanks

 With state access, we can modify balances, read/write the storage of other contracts, and could even hook into external storage outside of the bounds of the EVM’s merkle trie (note: this would come with repercussions for fast sync since part of the state would be moved off of the merkle trie). We can now write custom logic to make our own EVM. We can do more on Avalanche in Solidity than on Ethereum!

## The Process

> We will first create a Solidity interface that our precompile will implement.  Then we will use the precompile tool to autogenerate functions and fill out the rest. We're not done yet! We will then have to update a few more places within the EVM. Some of this work involves assigning a precompile address, adding the precompile to the list of EVM precompiles, and finally enabling the precompile. Now we can see our functions in action as we write another solidity smart contract that interacts with our precompile. Lastly, we will write some tests to make sure everything works as promised. 

Reviewer 2: ```suggestion
We will first create a Solidity interface that our precompile will implement.  Then we will use the precompile tool to autogenerate functions and fill out the rest. We're not done yet! We will then have to update a few more places within the Subnet-EVM. Some of this work involves assigning a precompile address, adding the precompile to the list of Subnet-EVM precompiles, and finally enabling the precompile. Now we can see our functions in action as we write another solidity smart contract that interacts with our precompile. Lastly, we will write some tests to make sure everything works as promised. 
```

Reviewer 2: We should use Subnet-EVM instead of EVM where possible

## Step 8: Create Genesis

We can move our genesis file we created in the last step to  `tests/e2e/genesis/`.

```cp /tmp/subnet-evm-genesis.json  tests/e2e/genesis/hello_world.json``` 

## Step 9: Add E2E tests

In `tests/e2e/solidity/suites.go` we can now write our first e2e test!
It's another nice copy and paste situation. 

``` go
ginkgo.It("hello world", func() {
		err := startSubnet("./tests/e2e/genesis/hello_world.json")
		gomega.Expect(err).Should(gomega.BeNil())
		running := runner.IsRunnerUp()
		gomega.Expect(running).Should(gomega.BeTrue())
		runHardhatTests("./test/ExampleHelloWorld.ts")
		stopSubnet()
		running = runner.IsRunnerUp()
		gomega.Expect(running).Should(gomega.BeFalse())
	})

	// ADD YOUR PRECOMPILE HERE
	/*
			ginkgo.It("your precompile", func() {
			err := startSubnet("./tests/e2e/genesis/{your_precompile}.json")
			gomega.Expect(err).Should(gomega.BeNil())
			running := runner.IsRunnerUp()
			gomega.Expect(running).Should(gomega.BeTrue())
			runHardhatTests("./test/Example{YourPrecompile}Test.ts")
			stopSubnet()
			running = runner.IsRunnerUp()
			gomega.Expect(running).Should(gomega.BeFalse())
		})
	*/
```

> ## Step 10: Run E2E Test

Reviewer 2: if possible can we add a step to show how to call this ExampleHelloWorld contract from Remix and show results?

# Stateful Precompile Generation Tutorial
In this tutorial,  we are going to walkthrough how we can generate a stateful precompile from scratch. Before we start, let's brush up on what a precompile is, what a stateful precompile is, and why this is extremely useful. 

## Precompiled Contracts

> Precompiles were introduced to Ethereum as a way to solve the problem of allowing complex cryptographic computations to be usable in the EVM without having to deal with EVM overhead. The following precompiles are currently included: ecrecover, sha256, blake2f, ripemd-160, Bn256Add, Bn256Mul, Bn256Pairing, the identity function, and modular exponentiation.

Reviewer: ```suggestion
Ethereum uses precompiles to efficiently implement cryptographic primitives within the EVM instead of re-implementing the same primitives in Solidity. The following precompiles are currently included: ecrecover, sha256, blake2f, ripemd-160, Bn256Add, Bn256Mul, Bn256Pairing, the identity function, and modular exponentiation.
```

Author: thanks

```

Smart contracts in solidity are compiled and converted into bytecode when they are first deployed. They are then stored on the blockchain and an address (usually known as the contract address) is assigned to it. When a user calls a function from a smart contract, it goes through the `CALL` function in the EVM. It takes in the caller address, the contract address, the input (function’s signature (truncated to the first leading four bytes) followed by the packed arguments data), gas, and value (native token). The function selector from the input lets the EVM know where to start from in the bytecode of the smart contract. It then executes a series of instructions (EVM opcodes) and returns the result. 

When a precompile function is called, it still goes through the `CALL` function in the EVM. However, it works a little differently. The EVM checks if the address is a precompile address from the mapping list and if so redirects to the precompile function. 

``` go
  if p := precompiles[addr]; p != nil {
    return RunPrecompiledContract(p, input, contract)

>   }

Reviewer: This is not what the actual code looks like, so can we remove this example?

Author: This is how normal precompiles look like in Ethereum

Author: I was trying to show how ethereum does it.

Precompiles provide complex library functions that are commonly used in smart contracts and do not use EVM opcodes which makes execution faster and gas costs lower.

## Stateful Precompiled Contracts

A stateful precompile allows us to add even more functionality and customization to the EVM. It builds on a precompile in that it adds state access. 

A stateful precompile follows this interface. 
``` go
// StatefulPrecompiledContract is the interface for executing a precompiled contract
type StatefulPrecompiledContract interface {
	// Run executes the precompiled contract.
	Run(accessibleState PrecompileAccessibleState, caller common.Address, addr  common.Address, input []byte, suppliedGas uint64, readOnly bool) (ret []byte, remainingGas uint64, err error)

}
```

Notice the most important difference between the stateful precompile and precompile interface. We now inject state access to the `Run` function. Precompiles only took in a single byte slice as input. However, stateful precompile functions have complete access to the EVM state, and can be used to implement a much wider range of functionalities.

## Why This is Useful

>  With state access, we can modify balances, read/write the storage of other contracts, and could even hook into external storage outside of the bounds of the EVM’s merkle trie (note: this would come with repercussions for fast sync since part of the state would be moved off of the merkle trie). We can now write custom logic to make our own EVM. We can do more on Avalanche in Solidity than on Ethereum!

Reviewer: ```suggestion
 With state access, we can modify balances, read/write the storage of other contracts, and could even hook into external storage outside of the bounds of the EVM’s merkle trie (note: this would come with repercussions for state sync since part of the state would be moved off of the merkle trie). We can now write custom logic to make our own EVM.
```

Author: thanks

## Tutorial

We will first start off by creating the Solidity interface that we want our precompile to implement. This will be the HelloWorld Interface. It will have two simple functions, `sayHello` and `setGreeting`. These two functions will demonstrate the getting and setting respectively of a value using state access. 

We will place the interface in `./contract-examples/contracts`

``` sol
// (c) 2022-2023, Ava Labs, Inc. All rights reserved.
// See the file LICENSE for licensing terms.

// SPDX-License-Identifier: MIT

pragma solidity >=0.8.0;

interface IHelloWorld {
 // sayHello returns the string located at [key]
  function sayHello() external returns (string calldata);

// setGreeting sets the string located at [key]
  function setGreeting(string calldata response) external;
}
```

Now we have an interface that our precompile can implement!
Let's create an [abi](https://docs.soliditylang.org/en/v0.8.13/abi-spec.html#:~:text=Contract%20ABI%20Specification-,Basic%20Design,as%20described%20in%20this%20specification.) of our solidity code.

> In the same `./contract-examples/contracts` directory, let's [download solc](https://docs.soliditylang.org/en/v0.8.9/installing-solidity.html) and run

Reviewer: Can we switch from saying "in the same directory" to providing the full set of commands for a user to use?

Author: Do you want me to explicitly write the solc commands here? I feel like that's not in the scope of this tutorial.

> ```

Reviewer: ```suggestion
```bash
```

Author: thanks

solc --abi IHelloWorld.sol -o .
```

This spits out the abi code!

IHelloWorld.abi

``` json
[{"inputs":[],"name":"sayHello","outputs":[{"internalType":"string","name":"","type":"string"}],"stateMutability":"nonpayable","type":"function"},{"inputs":[{"internalType":"string","name":"recipient","type":"string"}],"name":"setGreeting","outputs":[],"stateMutability":"nonpayable","type":"function"}]
```

>  The precompile gen tool needs named outputs for the abi. Let's modify the abi and name our output `result`.

Reviewer: Instead of generating the ABI twice, can we just specify that the output arguments require names?

Reviewer: Actually, can we update the interface that we are asking them to copy and paste to include the names and add a note to tell the users that this is a requirement in order to generate the precompile template.

Author: Yup!

``` json
[{"inputs":[],"name":"sayHello","outputs":[{"internalType":"string","name":"result","type":"string"}],"stateMutability":"nonpayable","type":"function"},{"inputs":[{"internalType":"string","name":"recipient","type":"string"}],"name":"setGreeting","outputs":[],"stateMutability":"nonpayable","type":"function"}]

> ```

Reviewer: There's a missing step here to actually put the ABI code into a correct location to be used in the next step.

Author: I specify the correct location in the paragraph above. In the same `./contract-examples/contracts` directory, let's [download solc](https://docs.soliditylang.org/en/v0.8.9/installing-solidity.html) and run

``` go
/* General guidelines for precompile development:
1- Read the comment and set a suitable contract address in precompile/params.go. E.g:
	HelloWorldAddress = common.HexToAddress("ASUITABLEHEXADDRESS")
2- Set gas costs here
3- It is recommended to only modify code in the highlighted areas marked with "CUSTOM CODE STARTS HERE". Modifying code outside of these areas should be done with caution and with a deep understanding of how these changes may impact the EVM.
Typically, custom codes are required in only those areas.
4- Add your upgradable config in params/precompile_config.go
5- Add your precompile upgrade in params/config.go
6- Add your solidity interface and test contract to contract-examples/contracts
7- Write solidity tests for your precompile in contract-examples/test
8- Create your genesis with your precompile enabled in tests/e2e/genesis/
9- Create e2e test for your solidity test in tests/e2e/solidity/suites.go
10- Run your e2e precompile Solidity tests with 'E2E=true ./scripts/run.sh'
*/
```

## Step 1: Set Contract Address

> In `./precompile/params.go` let's set a precompile address. We can cut the

Reviewer: Can we make this more clear by stating exactly where to find the original address that needs to be moved and then stating that it needs to be moved to in `./precompile/params.go` ?

Reviewer: This starts out with `./precompile/params.go` but I think it would be more clear if we start with the location in the generated precompile file.

Author: fixed

address from the var declaration block and remove it from the precompile. 

![](2022-09-01-22-46-00.png)
![](2022-08-24-16-45-48.png)

Now when subnet-evm sees the `HelloWorldAddress` as input when executing [`CALL`](../core/vm/evm.go#L222), [`STATICCALL`](../core/vm/evm.go#L401), [`DELEGATECALL`](core/vm/evm.go#L362), [`CALLCODE`](core/vm/evm.go#L311), it can [run the precompile](https://github.com/ava-labs/subnet-evm/blob/master/core/vm/evm.go#L271-L272) if the precompile is enabled.

## Step 2: Set Gas Costs

Set up gas costs. In `precompile/params.go` we have `writeGasCostPerSlot` and `readGasCostPerSlot`. This is a good starting point for estimating gas costs. 

``` go
// Gas costs for stateful precompiles
const (
	writeGasCostPerSlot = 20_000
	readGasCostPerSlot  = 5_000
)
```

For example, we will be setting and getting our greeting in one slot so we can define the gas costs as follows. 

``` go
	SayHelloGasCost uint64    = 20_000 

> 	SetGreetingGasGost uint64 = 5000

Author: Yup thanks for catching this

```

**Example:** 
The sha256 precompile computes gas with the following equation
``` go
// This method does not require any overflow checking as the input size gas costs
// required for anything significant is so high it's impossible to pay for.
func (c *sha256hash) RequiredGas(input []byte) uint64 {
	return uint64(len(input)+31)/32*params.Sha256PerWordGas + params.Sha256BaseGas
}
```

## Step 3: Add Custom Code

Ok time to `CTRL F` throughout the file with `CUSTOM CODE STARTS HERE` to find the areas in the precompile that we need to modify. 

![](2022-09-01-22-48-26.png)

If we use any of the following imports we can remove the reference imports.
We can come back if needed to remove. Next we see this in `Equals()`.

> 

Reviewer: If the imports are unused then we can remove this line as well as the import itself.

We include this in the template so that the template can simply include all of the imports that it might need and avoid generating a file that may have un-needed imports which would result in a compiler warning.

Author: thanks

![](2022-09-01-22-51-30.png)

We can skip this step since our HelloWorldConfig struct looks like this.

```
// HelloWorldConfig implements the StatefulPrecompileConfig
// interface while adding in the HelloWorld specific precompile address.
type HelloWorldConfig struct {
	UpgradeableConfig
}
```

**Optional Note** 

If our HelloWorldConfig wrapped another config in its struct to implement the StatefulPrecompileConfig
like so, 

``` go
// HelloWorldConfig implements the StatefulPrecompileConfig
// interface while adding in the IHelloWorld specific precompile address.

type HelloWorldConfig struct {
	UpgradeableConfig
    AllowListConfig
}
```

We would have to modify the `Equal()` function as follows: 

![](2022-09-01-22-54-22.png)

The next place we see the `CUSTOM CODE STARTS HERE` is in `Configure()`

> Let's set it up. 

Author: I believe some of the comments in Configure explain it, but I'll add more of a description.

``` go
// Configure configures [state] with the initial configuration.
func (c *HelloWorldConfig) Configure(_ ChainConfig, state StateDB, _ BlockContext) {
	// CUSTOM CODE STARTS HERE
	// This will be called in the first block where HelloWorld stateful precompile is enabled.
	// 1) If BlockTimestamp is nil, this will not be called
	// 2) If BlockTimestamp is 0, this will be called while setting up the genesis block
	// 3) If BlockTimestamp is 1000, this will be called while processing the first block
	// whose timestamp is >= 1000
	//
	// Set the initial value under [common.BytesToHash([]byte("storageKey")] to "Hello World!"
	res := common.LeftPadBytes([]byte("Hello World!"), common.HashLength)
	state.SetState(HelloWorldAddress, common.BytesToHash([]byte("storageKey")), common.BytesToHash(res))
}
```

> We also see a `Verify()` function. We can leave this as is for now. 

Reviewer: Can we provide an explanation of how `Verify()` works here (there's an explanation in `./precompile/stateful_precompile_config.go` and explain that we leave this as a no-op because there is no invalid configuration for the `HelloWorldConfig`

Reviewer: Can we explain why there is no verification needed within this function and then ask the user to delete the CUSTOM CODE comment and provide a copy pastable version of this code with the comment removed that the user can replace the code with?

Author: Yes thanks

``` go
// Verify tries to verify HelloWorldConfig and returns an error accordingly.
func (c *HelloWorldConfig) Verify() error {

	// CUSTOM CODE STARTS HERE
	// Add your own custom verify code for HelloWorldConfig here
	// and return an error accordingly
	return nil
}
```

> Next place to modify is in our `sayHello()` function.

Reviewer: Can we provide some description of what the sayHello and setGreeting functions are? We should at least explain that these are the execution functions that are generated for each function we defined in the solidity interface and will be executed when that interface is called.

Author: Yup absolutely

		return nil, remainingGas, err
	}

	// Return the packed output and the remaining gas
	return packedOutput, remainingGas, nil
}
```

Finally we can modify our `setGreeting()` function. 

``` go
func setGreeting(accessibleState PrecompileAccessibleState, caller common.Address, addr common.Address, input []byte, suppliedGas uint64, readOnly bool) (ret []byte, remainingGas uint64, err error) {
	if remainingGas, err = deductGas(suppliedGas, SetGreetingGasCost); err != nil {
		return nil, 0, err
	}
	if readOnly {
		return nil, remainingGas, vmerrs.ErrWriteProtection
	}
	// attempts to unpack [input] into the arguments to the SetGreetingInput.
	// Assumes that [input] does not include selector
	// You can use unpacked [inputStruct] variable in your code
    inputStr, err := UnpackSetGreetingInput(input)
	if err != nil {
		return nil, remainingGas, err
	}

	// CUSTOM CODE STARTS HERE
    // check if the input string is longer than 32 bytes
    if len(inputStr) > 32 {
      return nil, 0, errors.New("input string is longer than 32 bytes")
    }

	// setGreeting is the execution function "SetGreeting(name string)"
    // and sets the storageKey in the string returned by hello world

>   

Reviewer: Can we fix the indenting here?

Author: Unfortunately if I fix the indenting there, it messes up the actual indenting in the codeblock in the actual markdown file

Author: Actually I did some work around to fix it thanks

    res := common.LeftPadBytes([]byte(inputStr), common.HashLength)
	accessibleState.GetStateDB().SetState(HelloWorldAddress, common. BytesToHash([]byte("storageKey")), common.BytesToHash(res))

	// this function does not return an output, leave this one as is
	packedOutput := []byte{}

	// Return the packed output and the remaining gas
	return packedOutput, remainingGas, nil
}
```

## Step 4: Add Upgradable Config

Let's now modify `params/precompile_config.go`. We can `CTRL F` for `ADD YOUR PRECOMPILE HERE`. 

> Let's add our key. Note that there are two places to in the image below. 

Author: Thanks

![](2022-09-01-23-20-53.png)

> Let's add our precompile config to `PrecompileUpgrade`

Reviewer: We should make each of these instructions throughout the tutorial as specific as possible

Author: Yup

![](2022-09-01-23-19-42.png)

> Finally, we can add a getter function. 

Reviewer: We should explain each of the pieces of the getter function since it is a very convoluted getter function. The user needs to update the comment, function signature, return type, key parameter to `getActivePrecompileConfig`, and the casted type to return in the body of the if condition.

![](2022-09-01-23-23-20.png)

> Done! All we had to do was follow the comments.

Reviewer: Can we modify these to be code snippets instead of screenshots so that they are copy-pastable?

Author: Yup

> Done! All we had to do was follow the comments.

Reviewer: This is missing one more place that we need to modify within `GetActivePrecompiles`

Author: Thanks for catching that

## Step 5: Add Precompile Upgrade

Let's add our precompile upgrade in `params/config.go`. We can `CTRL F` for `ADD YOUR PRECOMPILE HERE`. 

Let's add the bool to check if our precompile is enabled.  

![](2022-09-01-23-35-47.png)

We can now add it to the Avalanche rules. 

![](2022-09-01-23-36-56.png)

> Lastly, we can add `IsHelloWorld` which checks if we are equal or greater than the fork `blockTimestamp`. 

Reviewer: We should change the order of steps here. The `IsHelloWorld` function is used by the previous step, so we should first add this function, then add it to rules, and then set it in the `AvalancheRules(...)` function.

Author: thanks

// each of the possible stateful precompile types that can be activated
// as a network upgrade.
type PrecompileUpgrade struct {
	ContractDeployerAllowListConfig *precompile.ContractDeployerAllowListConfig `json:"contractDeployerAllowListConfig,omitempty"` // Config for the contract deployer allow list precompile
	ContractNativeMinterConfig      *precompile.ContractNativeMinterConfig      `json:"contractNativeMinterConfig,omitempty"`      // Config for the native minter precompile
	TxAllowListConfig               *precompile.TxAllowListConfig               `json:"txAllowListConfig,omitempty"`               // Config for the tx allow list precompile
	FeeManagerConfig                *precompile.FeeConfigManagerConfig          `json:"feeManagerConfig,omitempty"`                // Config for the fee manager precompile
	HelloWorldConfig                *precompile.HelloWorldConfig                `json:"helloWorldConfig,omitempty"`
	// ADD YOUR PRECOMPILE HERE
	// {YourPrecompile}Config  *precompile.{YourPrecompile}Config `json:"{yourPrecompile}Config,omitempty"`
}
```

Now we can get the network up and running. 
Open some terminal tabs and enter the following commands.

> Here's some more information on [avalanche-network-runner](https://docs.avax.network/subnets/network-runner) and how to download it. 

Reviewer: Can we move this to a requirements section at the top of the tutorial?

Author: I don't believe it belongs there. The links at the top are more about evm and precompiles knowledge. This link is about a developer tool to start up a local network. It makes sense to leave it here in the flow of the tutorial IMO.

``` go
// Start the server 
avalanche-network-runner server \
--log-level debug \
--port=":8080" \
--grpc-gateway-port=":8081"

// In the root of the repo, run this to get the latest subnetevm binary 
./scripts/build.sh

// Set paths

> export AVALANCHEGO_EXEC_PATH="${HOME}/go/src/avalanchego/build/avalanchego"

Reviewer: I think the better version of this would be to use `GOPATH/src/...` instead of assuming the user has set up their `GOPATH` directly within the home directory.

Reviewer: Actually this is missing `github.com/ava-labs` before `avalanchego`.

Author: ah thanks.

export AVALANCHEGO_PLUGIN_PATH="${HOME}/go/src/github.com/ava-labs/avalanchego/build/plugins"
 
// Spin up some nodes that run the latest version of subnet evm 
  avalanche-network-runner control start \
  --log-level debug \
  --endpoint="0.0.0.0:8080" \
  --number-of-nodes=5 \
  --avalanchego-path ${AVALANCHEGO_EXEC_PATH} \

>   --plugin-dir ${AVALANCHEGO_PLUGIN_PATH} \

Reviewer: Is the preferred way of using the network runner to export these variables and pass them in this way cc @dasconnor ?

>   --blockchain-specs '[{"vm_name": "subnetevm", "genesis": "/tmp/subnet-evm-genesis.json"}]'

Reviewer: Can we clean this section up a little bit and clearly separate each step?

Author: Yup thanks

> ```

Reviewer: This piece is pretty messy. We should break this down more clearly

Author: fixed thank you

If the network startup is successful then you should see something like this.
```
[blockchain RPC for "srEXiWaHuhNyGwPUi444Tu47ZEDwxTWrbQiuD7FmgSAQ6X7Dy"] "http://127.0.0.1:9650/ext/bc/2jDWMrF9yKK8gZfJaaaSfACKeMasiNgHmuZip5mWxUfhKaYoEU"
[blockchain RPC for "srEXiWaHuhNyGwPUi444Tu47ZEDwxTWrbQiuD7FmgSAQ6X7Dy"] "http://127.0.0.1:9652/ext/bc/2jDWMrF9yKK8gZfJaaaSfACKeMasiNgHmuZip5mWxUfhKaYoEU"
[blockchain RPC for "srEXiWaHuhNyGwPUi444Tu47ZEDwxTWrbQiuD7FmgSAQ6X7Dy"] "http://127.0.0.1:9654/ext/bc/2jDWMrF9yKK8gZfJaaaSfACKeMasiNgHmuZip5mWxUfhKaYoEU"
[blockchain RPC for "srEXiWaHuhNyGwPUi444Tu47ZEDwxTWrbQiuD7FmgSAQ6X7Dy"] "http://127.0.0.1:9656/ext/bc/2jDWMrF9yKK8gZfJaaaSfACKeMasiNgHmuZip5mWxUfhKaYoEU"
[blockchain RPC for "srEXiWaHuhNyGwPUi444Tu47ZEDwxTWrbQiuD7FmgSAQ6X7Dy"] "http://127.0.0.1:9658/ext/bc/2jDWMrF9yKK8gZfJaaaSfACKeMasiNgHmuZip5mWxUfhKaYoEU"
```

> Sweet! Now we have blockchain rpcs that can be used to talk to the network!

Reviewer: ```suggestion
Sweet! Now we have blockchain RPCs that can be used to talk to the network!
```

Author: thanks
